### ISSUE TITLE
Connecting to cluster does not work

### ISSUE THREAD

**Issue by pkruithof** [prooph/event-store-client#34]
Steps to reproduce:

```
cd docker/local-3-node-dns-cluster
docker-compose up -d
```

I have to add the following to `docker/unittest/docker-compose.yml`:

```
services:
  php:
[...]
#    depends_on:
#      - eventstore

networks:
  clusternetwork:
    external:
      name: local-3-node-dns-cluster_clusternetwork
```

I have to adjust the example as it gives the following exception right now:

```
/app/examples # php demo-async-dns-cluster.php 

Fatal error: Uncaught Prooph\EventStoreClient\Exception\InvalidArgumentException: Must specify uri or gossip seeds in /app/src/EventStoreAsyncConnectionFactory.php:138
Stack trace:
#0 /app/src/EventStoreAsyncConnectionFactory.php(163): Prooph\EventStoreClient\EventStoreAsyncConnectionFactory::createFromUri(NULL, Object(Prooph\EventStoreClient\ConnectionSettings), 'dns-cluster-con...')
#1 /app/examples/demo-async-dns-cluster.php(27): Prooph\EventStoreClient\EventStoreAsyncConnectionFactory::createFromSettings(Object(Prooph\EventStoreClient\ConnectionSettings), 'dns-cluster-con...')
#2 [internal function]: Prooph\EventStoreClient\{closure}('a', NULL)
#3 /app/vendor/amphp/amp/lib/Coroutine.php(39): Generator->current()
#4 /app/vendor/amphp/amp/lib/Loop/Driver.php(119): Amp\Coroutine->__construct(Object(Generator))
#5 /app/vendor/amphp/amp/lib/Loop/Driver.php(70): Amp\Loop\Driver->tick()
#6 /app/vendor/amphp/amp/lib/Loop.php(76): Amp\Loop\Driver->run()
#7 /app/examples/demo-async-dns-cluster.php(102): Amp\Loop::run(Object(Clo in /app/src/EventStoreAsyncConnectionFactory.php on line 138
```

After changing it to this:

```
    $builder = new ConnectionSettingsBuilder();
    $builder->setClusterDns('escluster.net');
    $builder->setClusterGossipPort(2112); # not 2113 as the example has right now

    $connection = EventStoreAsyncConnectionFactory::createFromConnectionString(
        'ConnectTo=discover://escluster.net:2113',
        $builder->build(),
        'dns-cluster-connection'
    );
```

the output is this:

```
/app # php examples/demo-async-dns-cluster.php 
connection closed

Fatal error: Uncaught Prooph\EventStoreClient\Exception\CannotEstablishConnectionException: Cannot resolve target end point in /app/src/Internal/EventStoreConnectionLogicHandler.php:245
Stack trace:
#0 /app/vendor/amphp/amp/lib/Failure.php(26): Prooph\EventStoreClient\Internal\EventStoreConnectionLogicHandler->Prooph\EventStoreClient\Internal\{closure}(Object(Prooph\EventStoreClient\Exception\ClusterException), NULL)
#1 /app/vendor/amphp/amp/lib/Internal/Placeholder.php(122): Amp\Failure->onResolve(Object(Closure))
#2 /app/vendor/amphp/amp/lib/Internal/Placeholder.php(151): Amp\Coroutine->resolve(Object(Amp\Failure))
#3 /app/vendor/amphp/amp/lib/Coroutine.php(93): Amp\Coroutine->fail(Object(Prooph\EventStoreClient\Exception\ClusterException))
#4 /app/vendor/amphp/amp/lib/Internal/Placeholder.php(127): Amp\Coroutine->Amp\{closure}(Object(Prooph\EventStoreClient\Exception\ClusterException), NULL)
#5 /app/vendor/amphp/amp/lib/Delayed.php(20): Amp\Delayed->resolve(NULL)
#6 /app/vendor/amphp/amp/lib/Loop/NativeDriver.php(101): in /app/src/Internal/EventStoreConnectionLogicHandler.php on line 245
```

**Comment by prolic** [comment]
Ok, so the problem is that it will always try to connect to external ip (not internal ip). In case of a docker container, your external ip is always 127.0.0.1. This will obviously then make no connection to the event store.

I'm working on the feature of allowing internal ips as well.

**Comment by prolic** [comment]
A damn it, that's not possible, event-store itself has to support connecting to internal ip addresses from client, first.

Gossip port should be 2112, you're right. But without public ip (not 127.0.0.1 bullshit) you cannot run the event-store cluster in docker.

I'll recheck all demo scripts asap.

**Comment by prolic** [comment]
@pkruithof You could try setting EVENTSTORE_EXT_IP in your docker-compose file, not sure if that works

**Comment by pkruithof** [comment]
Some new findings:

Connecting to the cluster does work when running the demo locally (outside of Docker). Applying #35 helps in this matter.

The main issue that led me into this rabbit hole is consuming from a persistent subscription. What I have found is that when connecting to a persistent subscription, a cluster will return a `NotHandled: NotMaster` package, which should cause the client to reconnect to the given endpoint. This does happen when calling `$connection->createPersistentSubscriptionAsync()` for example, which is why the demo works, because it does this:

```php
# dummy code
yield $connection->connectAsync();
yield $connection->createPersistentSubscriptionAsync()
yield $connection->connectToPersistentSubscriptionAsync();
```

without the `createPersistentSubscriptionAsync` call, this will fail immediately, see this log: https://gist.github.com/pkruithof/bfc57ece5b4cf7d7af4155941fa05b74

When it's added (or alternatively `$connection->updatePersistentSubscriptionAsync()` if the subscription already exists), the subscription is connected to, but then dropped because of a double resolved promise:

```
[2018-11-16 15:55:56] event-store-client.debug: EventStoreNodeConnection 'dns-cluster-connection': TcpConnectionError connId 1fa54da21d2845d880a863b2f706248c, exception Promise has already been resolved, define environment variable AMP_DEBUG or const AMP_DEBUG = true and enable assertions for a stacktrace of the previous resolution.
```

I placed debug statements at the points where promises were resolved, and got this in my console:

```
[2018-11-16 15:56:23] event-store-client.info: Discovering: found best choice [127.0.0.1:1113, n/a] (Slave)
[2018-11-16 15:56:23] event-store-client.info: Discovering attempt 1/10 successful: best candidate is [127.0.0.1:1113, n/a]
[2018-11-16 15:56:23] event-store-client.debug: TcpPackageConnection: connected to [127.0.0.1:1113, 8f671b465c2f42f7b48bdb9df08984ce]
connected
[2018-11-16 15:56:23] event-store-client.debug: TcpPackageConnection: connected to [127.0.0.1:1123, 570958a83f5947c780db2b11576f16cf]
connected
RESOLVE OPERATION Prooph\EventStoreClient\ClientOperations\AppendToStreamOperation
RESOLVE OPERATION Prooph\EventStoreClient\ClientOperations\UpdatePersistentSubscriptionOperation
RESOLVE CONFIRM SUBSCRIPTION
RESOLVE START SUBSCRIPTION
incoming event: 581@foo-bar
data: 4f49cc3fc987eda24cfa9b27
RESOLVE OPERATION Prooph\EventStoreClient\ClientOperations\AppendToStreamOperation
dropped with reason: UserInitiated
RESOLVE OPERATION Prooph\EventStoreClient\ClientOperations\UpdatePersistentSubscriptionOperation
connection closed
```

The demo script is located here: https://gist.github.com/pkruithof/ac666c7a3a5d1e9087eed4037320fbba

So, TL;DR:
1. Reconnecting is not being handled for `connectToPersistentSubscriptionAsync` calls
2. Sometimes operations are resolved more than once

Hope this is enough for you to go on.

**Comment by pkruithof** [comment]
Most of the issues here are fixed now. I created a separate issue for the remaining issue: #37
